internal/app/document_store.go

```diff
@@ -0,0 +1,109 @@
+package app
+
+import (
+	"context"
+	"errors"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/dynamodb"
+	"github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
+	"github.com/ministryofjustice/opg-modernising-lpa/internal/dynamo"
+	"github.com/ministryofjustice/opg-modernising-lpa/internal/page"
+)
+
+type documentStore struct {
+	dynamoClient DynamoClient
+	s3Client     S3Client
+}
+
+func NewDocumentStore(dynamoClient DynamoClient, s3Client S3Client) DocumentStore {
+	return documentStore{dynamoClient: dynamoClient, s3Client: s3Client}
+}
+
+func (s documentStore) GetAll(ctx context.Context) (page.Documents, error) {
+	data, err := page.SessionDataFromContext(ctx)
+	if err != nil {
+		return []page.Document{}, err
+	}
+
+	if data.LpaID == "" {
+		return []page.Document{}, errors.New("documentStore.GetAll requires LpaID")
+	}
+
+	var ds []page.Document
+	if err := s.dynamoClient.AllByPartialSk(ctx, lpaKey(data.LpaID), DocumentKey(""), &ds); err != nil {
+		if errors.Is(err, dynamo.NotFoundError{}) {
+			return []page.Document{}, nil
+		}
+
+		return []page.Document{}, err
+	}
+
+	return ds, nil
+}
+
+func (s documentStore) UpdateScanResults(ctx context.Context, PK, SK string, virusDetected bool) error {
+	input := &dynamodb.UpdateItemInput{
+		Key: map[string]types.AttributeValue{
+			"PK": &types.AttributeValueMemberS{Value: PK},
+			"SK": &types.AttributeValueMemberS{Value: SK},
+		},
+		ExpressionAttributeValues: map[string]types.AttributeValue{
+			":virusDetected": &types.AttributeValueMemberBOOL{Value: virusDetected},
+			":scanned":       &types.AttributeValueMemberBOOL{Value: true},
+		},
+		UpdateExpression: aws.String("set VirusDetected = :virusDetected, Scanned = :scanned"),
+	}
+
+	return s.dynamoClient.Update(ctx, input)
+}
+
+func (s documentStore) Put(ctx context.Context, document page.Document, data []byte) error {
+	if data != nil {
+		if err := s.s3Client.PutObject(ctx, document.Key, data); err != nil {
+			return err
+		}
+	}
+
+	return s.dynamoClient.Put(ctx, document)
+}
+
+func (s documentStore) DeleteInfectedDocuments(ctx context.Context, documents page.Documents) error {
+	var dynamoKeys []dynamo.Key
+	var s3Keys []string
+
+	for _, d := range documents {
+		if d.VirusDetected {
+			dynamoKeys = append(dynamoKeys, dynamo.Key{
+				PK: d.PK,
+				SK: d.SK,
+			})
+			s3Keys = append(s3Keys, d.Key)
+		}
+	}
+
+	if len(dynamoKeys) == 0 {
+		return nil
+	}
+
+	if err := s.s3Client.DeleteObjects(ctx, s3Keys); err != nil {
+		return err
+	}
+
+	return s.dynamoClient.DeleteKeys(ctx, dynamoKeys)
+}
+
+func (s documentStore) Delete(ctx context.Context, document page.Document) error {
+	if err := s.s3Client.DeleteObject(ctx, document.Key); err != nil {
+		return err
+	}
+
+	return s.dynamoClient.DeleteOne(ctx, dynamo.Key{
```

Comment:
why do the `dynamo.Key` thing here, could just pass `pk, sk string` as it only gets decomposed

Reply:
Fell into the trap of copying the signature for `DeleteKeys()` and dropping the slice - will update 👍